packages/git/src/browser/index.ts

```diff
@@ -0,0 +1,40 @@
+/********************************************************************************
+ * Copyright (C) 2018 Arm and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { FileStat } from '@theia/filesystem/lib/common';
+import { Git, Repository } from '../common';
+
+export const ScmWidgetFactory = Symbol('ScmWidgetFactory');
+
+export const GIT_WIDGET_FACTORY_ID = 'git';
+
+export interface ScmWidgetFactory {
+    readonly label: string;
+    readonly widgetId: string;
+
+    /**
+     * Given a directory FileStat populated with its children, determine if this is a repository
+     * of appropriate type.
+     *
+     * @param fileStat
+     */
+    isUnderSourceControl(fileStat: FileStat): boolean;
+
+    /**
+     * Resolves to an array of repositories discovered in the workspace given with the workspace root URI.
+     */
+    repositories(workspaceRootUri: string, options: Git.Options.Repositories): Promise<Repository[]>;
```

Comment:
I understood this is intended to abstract over different scms, but it uses git specific API.

Reply:
> I am really not sure if abstracting over SCMs is a good idea. I can see that mercurial is similar to Git but others are often not.

SVN and other non-distributed SCMs will have more differences than Mercurial, but I still see most of the components in the Git views being reusable for other SCMs.  I really don't see why it would not be a good idea to allow SCM implementations to re-use the components that apply and override the components that differ.  VS Code provides five extension points to be used by SCM implementations and I assume we would do the same.
  
> It's good to have the top level detection of what widget to show as the SCM widget, but that doesn't need to share concepts like repository or so.

You are right that Repository is in the Git namespace and is used for all SCMs.  This certainly could be cleaned up.  However Repository only has a single property 'localUri'.  All SCMs must result in local copy of the code and this is the URI of that local copy, so 'localUri' will apply to all SCMs.  Perhaps the Repository class was expected to expand so that it includes Git-specific properties.  (I had considered adding the SCM type to this class.)  I could share at the top level just the local URI, or I could create a base Repository class which can be used as is or extended by SCM providers.

> Also what is the use case of having multiple scm containers in one view?

You are right that we could provide a separate view for each SCM.  However that would result in multiple tabs, multiple views to be shown or hidden in the 'view' menu.  There would also be multiple components in the status bar, one to select from all the Git repositories, another to select from all the Mercurial repositories.  This would rapidly fill up the status bar.  Surely it is a better user experience if there is a single SCM view, and a single list from which the user can select a repository.  The single view wil, of course, react to the contents.  So, for example, if the repository is Mercurial then there will be no 'staged' section.  If the repository is SVN then there will be further differences.  Maybe I am missing something but that is the expectation of our users.  VS Code has a single view for all SCMs.